src/pmu.rs

```diff
@@ -0,0 +1,301 @@
+}
+
+impl PMUExt for PMU {
+    fn load_default_programs(&self) {
+        unsafe {
+            for i in 0..8 {
+                self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+                self.pmusleeppm[i].write(|w| w.bits(DEFAULT_SLEEP_PROGRAM[i]));
+
+                self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+                self.pmuwakepm[i].write(|w| w.bits(DEFAULT_WAKE_PROGRAM[i]));
+            }
+        }
+    }
+
+    fn sleep(self, sleep_time: u32) {
+        unsafe {
+            let rtc = RTC::ptr();
+
+            // set interrupt source to RTC enabled, each pmu register needs key set before write
+            self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+            self.pmuie.write(|w| w.rtc().set_bit().dwakeup().set_bit());
+            // set RTC clock scale to once per second for easy calculation
+            (*rtc).rtccfg.write(|w| w.enalways().set_bit().scale().bits(15));
+            // get current RTC clock value scaled
+            let rtc_now = (*rtc).rtcs.read().bits();
+            // set RTC clock comparator
+            (*rtc).rtccmp.write(|w| w.bits(rtc_now + sleep_time));
+            // go to sleep for sleep_time seconds, need to set pmukey here as well
+            self.pmukey.write(|w| w.bits(PMU_KEY_VAL));
+            self.pmusleep.write(|w| w.sleep().set_bit());
+        }
+    }
+
+    fn store_backup<UD>(&self, user_data: &UD) -> Result<(), BackupError>
+    where
+        UD: Sized,
+    {
+        unsafe {
+            let backup = BACKUP::ptr();
+            let ud_size = core::mem::size_of::<UD>();
+
+            if ud_size > (*backup).backup.len() * BACKUP_REGISTER_BYTES {
+                return Err(BackupError::DataTooLarge);
+            }
+
+            if ud_size % BACKUP_REGISTER_BYTES != 0 {
+                return Err(BackupError::DataSizeInvalid);
+            }
+
+            let reg_count = ud_size / BACKUP_REGISTER_BYTES;
+
+            let ptr = user_data as *const _;
+            let ptr_u32 = ptr as *const u32;
+            let sliced = core::slice::from_raw_parts(ptr_u32, reg_count);
+
+            for i in 0..sliced.len() {
+                (*backup).backup[i].write(|w| w.bits(sliced[i]));
+            }
+
+            Ok(())
+        }
+    }
+
+    fn restore_backup<UD>(&self, user_data: UD) -> Result<UD, BackupError>
+    where
+        UD: Sized,
+    {
+        unsafe {
+            let backup = BACKUP::ptr();
+            let ud_size = core::mem::size_of::<UD>();
+
+            if ud_size > (*backup).backup.len() * BACKUP_REGISTER_BYTES {
+                return Err(BackupError::DataTooLarge);
+            }
+
+            if ud_size % BACKUP_REGISTER_BYTES != 0 {
+                return Err(BackupError::DataSizeInvalid);
+            }
+
+            let reg_count = ud_size / BACKUP_REGISTER_BYTES;
+
+            let ptr = &user_data as *const _;
+            let ptr_u32 = ptr as *mut u32;
+            let sliced = core::slice::from_raw_parts_mut(ptr_u32, reg_count);
```

Comment:
I doubt this magic is OK because `user_data` is not `mut`. Unsafe-blocks are not just for persuading the compiler, they actually should be save from the outside perspective.
And you are trying to serialize arbitrary user data here. I'm pretty sure that in some cases you can't do this in a safe way. It's better to use `serde`-serializable types or plain arrays instead.

Reply:
Another question is why do you need `user_data` as argument? You can create it inside the function.

Reply:
1. I didn't want to bring external crates into this since it's a fairly minimal abstraction. As far as I can say this will work consistently. Rust guarantees internal type byte order on a single compile run. Since this isn't serialization I didn't want to go further. This is only meant to be used for a single run (e.g. not store and restore with a different program version which could actually break). If you feel strongly about this I could bring in serde to get something like a byte array serializer if they have it.

2. I didn't want to cause any inadvertant allocation. I know the compiler will ellide 99% of the cases, but it just seemed better if the user provided the memory.

Reply:
Well, how are you going to serialize a structure with a reference inside? What would you expect after restore?
I agree that adding a serde dependency is not a good solution, but maybe we can provide a way to use serde if the user wants to. For example, to provide a byte-slice interface for user to save their data into and then save this byte-slice into the PMU registers. In this case the user will be responsible for correct serialization of their data.